### PR TITLE
Register vacuum into transaction log

### PR DESCRIPTION
**Pull request by YannByron**
## Description

When execute `Vacuum`, this pr will help to track which files have been deleted, and commit this info to the `_delta_log` direcotry.

## How was this patch tested?

this patch can be tested by UT. Or check the vacuum commit info after execute `vacuum` command.

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

**Comment by YannByron** [comment]
@zsxwing @tdas can you have a chance to take a look? Thanks.

**Comment by YannByron** [comment]
gently ping @vkorukanti .

File: core/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala [hunk 1]
```diff
@@ -448,6 +448,25 @@ case class RemoveFile(
 }
 // scalastyle:on
 
+case class DeleteFile(
```
Comment (vkorukanti) [inline core/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala:451]: Why do you need a new Action? Can't we use the existing action `RemoveFile`?

File: core/src/main/scala/io/delta/tables/execution/VacuumTableCommand.scala [hunk 2]
```diff
@@ -35,10 +38,25 @@ case class VacuumTableCommand(
...
-  override val output: Seq[Attribute] =
+  override val output: Seq[Attribute] = if (dryRun) {
     Seq(AttributeReference("path", StringType, nullable = true)())
+  } else {
+    Seq(
+      AttributeReference("path", StringType, nullable = true)(),
+      AttributeReference("numVacuumedFiles", LongType, nullable = true)(),
+      AttributeReference("numVacuumedBytes", LongType, nullable = true)()
+    )
+  }
```
[the remark this answers is not in the document]
Comment (YannByron) [reply core/src/main/scala/io/delta/tables/execution/VacuumTableCommand.scala:52]: ok. it should be.

File: core/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala (hunk 1 above)
Comment (YannByron) [reply core/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala:451]: `RemoveFile` represents an ineffective file in the current snapshot, but still exists in FileSystem/Cloud Storage.
But `DeleteFile` represents an action that delete a file from FileSystem/Cloud Storage. I think there are different.

**Comment by YannByron** [comment]
gently ping @scottsand-db

**Comment by scottsand-db** [comment]
Hi @YannByron. I'd like to rethink at a higher level what the goal of this PR is. So you want to register into the delta log the exact files that were actually deleted by VACUUM. Why? Does this help with any sort of compliance? Can't you rely on the cloud store to provide those metrics/history for you?

What if instead we just provide some operation metrics?

Such as:

```scala
val VACUUM_START = Set(
  "numFilesToDelete", // number of files that will be deleted by vacuum
  "sizeOfDataToDelete" // total size in bytes of files that will be deleted by vacuum
)

val VACUUM_END = Set(
  "numDeletedFiles", // number of files deleted by vacuum
  "numVacuumedDirectories" // number of directories vacuumed
)
```

**Comment by scottsand-db** [comment]
Hi @YannByron - just following up here. Would be great if you could continue this work and/or respond to my feedback in the comment above.

**Comment by scottsand-db** [comment]
@YannByron I quickly sketched this up here: https://github.com/delta-io/delta/pull/1552. Want to help review?

**Pull request closed without merging**